File: framework/languages/node/e2e/RegistryMessaging/Server/Provider/Handlers/provider-handlers.ts
import { Injectable } from '@nestjs/common';
import type {
  ZLinkMessageFlowEvent,
  ZLinkMessageFlowObserver,
  ZLinkMessageContext,
  ZLinkRequestHandler,
  ZLinkRouteMessageContext,
  ZLinkRouteRequestHandler,
  ZLinkSendHandler
} from '@zlink-systems/framework';
import type {
  PayloadRes,
  PayloadReq,
  ProfileMsg,
  ProfileRes,
  ProfileReq,
  ScenarioRouteReq,
  ScenarioRouteRes
} from '../../../Shared/messages';
import { sha256Hex } from '../../../Shared/messages';
import { EvidenceStore } from '../Infrastructure/evidence-store';

@Injectable()
export class ProfileRequestHandler implements ZLinkRequestHandler<ProfileReq, ProfileRes> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: ProfileReq, context: ZLinkMessageContext): Promise<ProfileRes> {
    if (request.value === 'slow') {
      await delay(1000);
    }
    if (request.value.startsWith('rm-b3-transition-')) {
      this.evidence.add(`profile-request-start|rid=${this.evidence.rid}|value=${request.value}`);
      await delay(1000);
    }
    this.evidence.add(`profile-request|rid=${this.evidence.rid}|value=${request.value}|packet=${context.packetName}`);
    return { value: `profile:${request.value}`, providerRid: this.evidence.rid };
  }
}

@Injectable()
export class ProfileCommandHandler implements ZLinkSendHandler<ProfileMsg> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(command: ProfileMsg, context: ZLinkMessageContext): Promise<void> {
    if (command.commandId.startsWith('rm-c9-slow-')) {
      await delay(1000);
    }
    this.evidence.add(`profile-command|rid=${this.evidence.rid}|command=${command.commandId}|packet=${context.packetName}`);
  }
}

@Injectable()
export class PayloadRequestHandler implements ZLinkRequestHandler<PayloadReq, PayloadRes> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: PayloadReq, context: ZLinkMessageContext): Promise<PayloadRes> {
    const hash = sha256Hex(request.payload);
    this.evidence.add(
      `payload-request|rid=${this.evidence.rid}|marker=${request.marker}`
      + `|length=${request.payload.length}|sha256=${hash}|packet=${context.packetName}`
    );
    return { marker: request.marker, length: request.payload.length, sha256: hash };
  }
}

@Injectable()
export class RoutePingHandler implements ZLinkRouteRequestHandler<ScenarioRouteReq, ScenarioRouteRes> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: ScenarioRouteReq, context: ZLinkRouteMessageContext): Promise<ScenarioRouteRes> {
    const sourceRid = String(context.sourceNodeRid);
    this.evidence.add(`route-request|rid=${this.evidence.rid}|source=${sourceRid}|value=${request.value}`);
    return { value: `route:${request.value}`, providerRid: this.evidence.rid, sourceRid };
  }
}

@Injectable()
export class EvidenceDispatchErrorObserver implements ZLinkMessageFlowObserver {
  constructor(private readonly evidence: EvidenceStore) {}

  onMessageFlow(flow: ZLinkMessageFlowEvent): void {
    if (flow.outcome !== 'failed') {
      return;
    }
    this.evidence.add(
      'dispatch-error'
      + `|surface=${flow.surface}`
      + `|kind=${flow.messageKind}`
      + `|reason=${flow.reason ?? '<null>'}`
      + `|action=${flow.action ?? '<null>'}`
      + `|packet=${flow.packetName ?? '<null>'}`
      + `|channel=${flow.channelName ?? '<null>'}`
    );
  }
}

function delay(milliseconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

File: framework/languages/java/e2e-kotlin/RegistryMessaging/Server/Provider/src/main/kotlin/systems/zlink/e2e/kotlin/registrymessaging/provider/Handlers/ProviderHandlers.kt
package Handlers


import systems.zlink.e2e.kotlin.registrymessaging.provider.Handlers
import systems.zlink.e2e.kotlin.registrymessaging.provider.Infrastructure
import java.util.logging.Handler
import java.util.logging.LogRecord
import java.util.logging.Logger
import java.security.MessageDigest
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import java.util.HexFormat
import systems.zlink.e2e.kotlin.registrymessaging.provider.Infrastructure.EvidenceStore
import systems.zlink.e2e.kotlin.registrymessaging.shared.PayloadRes
import systems.zlink.e2e.kotlin.registrymessaging.shared.PayloadReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileMsg
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileRes
import systems.zlink.e2e.kotlin.registrymessaging.shared.ProfileReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.Contracts
import systems.zlink.e2e.kotlin.registrymessaging.shared.ScenarioRoutePingReq
import systems.zlink.e2e.kotlin.registrymessaging.shared.ScenarioRoutePingRes
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.channels.ZLinkRouteMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.kotlin.ZLinkSuspendingRouteRequestHandler
import systems.zlink.framework.kotlin.ZLinkSuspendingSendHandler
import kotlinx.coroutines.delay

@ZLinkHandlerGroup(Contracts.CHANNEL_HANDLER_GROUP)
class ProfileRequestHandler(
    private val evidence: EvidenceStore,
) : ZLinkSuspendingRequestHandler<ProfileReq, ProfileRes> {
    override suspend fun handle(request: ProfileReq, context: ZLinkMessageContext): ProfileRes {
        if (request.value == "slow") {
            delay(1000)
        }
        evidence.add(
            "profile-request|rid=${evidence.rid}|instance=${evidence.instanceId}" +
                "|value=${request.value}|packet=${context.packetName()}",
        )
        return ProfileRes("profile:${request.value}", evidence.rid, evidence.instanceId)
    }
}

@ZLinkHandlerGroup(Contracts.CHANNEL_HANDLER_GROUP)
class ProfileCommandHandler(
    private val evidence: EvidenceStore,
) : ZLinkSuspendingSendHandler<ProfileMsg> {
    override suspend fun handle(message: ProfileMsg, context: ZLinkMessageContext) {
        if (message.commandId.startsWith("slow")) {
            delay(1000)
        }
        evidence.add("profile-command|rid=${evidence.rid}|command=${message.commandId}|packet=${context.packetName()}")
    }
}

@ZLinkHandlerGroup(Contracts.CHANNEL_HANDLER_GROUP)
class PayloadRequestHandler(
    private val evidence: EvidenceStore,
) : ZLinkSuspendingRequestHandler<PayloadReq, PayloadRes> {
    override suspend fun handle(request: PayloadReq, context: ZLinkMessageContext): PayloadRes {
        val digest = MessageDigest.getInstance("SHA-256").digest(request.payload.toByteArray(Charsets.UTF_8))
        val hash = HexFormat.of().formatHex(digest).uppercase()
        evidence.add(
            "payload-request|rid=${evidence.rid}|marker=${request.marker}" +
                "|length=${request.payload.length}|sha256=$hash|packet=${context.packetName()}",
        )
        return PayloadRes(request.marker, request.payload.length, hash)
    }
}

@ZLinkHandlerGroup(Contracts.ROUTE_HANDLER_GROUP)
class RoutePayloadRequestHandler(
    private val evidence: EvidenceStore,
) : ZLinkSuspendingRouteRequestHandler<PayloadReq, PayloadRes> {
    override suspend fun handle(request: PayloadReq, context: ZLinkRouteMessageContext): PayloadRes {
        val digest = MessageDigest.getInstance("SHA-256").digest(request.payload.toByteArray(Charsets.UTF_8))
        val hash = HexFormat.of().formatHex(digest).uppercase()
        evidence.add(
            "payload-request|rid=${evidence.rid}|marker=${request.marker}" +
                "|length=${request.payload.length}|sha256=$hash|packet=${context.packetName()}",
        )
        return PayloadRes(request.marker, request.payload.length, hash)
    }
}

@ZLinkHandlerGroup(Contracts.ROUTE_HANDLER_GROUP)
class RoutePingHandler(
    private val evidence: EvidenceStore,
) : ZLinkSuspendingRouteRequestHandler<ScenarioRoutePingReq, ScenarioRoutePingRes> {
    override suspend fun handle(request: ScenarioRoutePingReq, context: ZLinkRouteMessageContext): ScenarioRoutePingRes {
        val source = context.sourceNodeRid().toString()
        evidence.add("route-request|rid=${evidence.rid}|source=$source|value=${request.value}")
        return ScenarioRoutePingRes("route:${request.value}", evidence.rid, source)
    }
}

class EvidenceDispatchErrorHandler(
    private val evidence: EvidenceStore,
) : Handler() {
    fun install() = Logger.getLogger(LOGGER_NAME).addHandler(this)

    override fun publish(record: LogRecord) {
        val fields = diagnosticsFields(record.message) ?: return
        if (fields["outcome"] != "ERROR") return
        evidence.add(
            "dispatch-error" +
                "|surface=${fields["surface"]}" +
                "|kind=${fields["kind"]}" +
                "|reason=${fields["reason"]}" +
                "|action=${fields["action"]}" +
                "|packet=${fields["packet"] ?: "<null>"}" +
                "|channel=${fields["channel"] ?: "<null>"}",
        )
    }

    override fun flush() = Unit
    override fun close() = Logger.getLogger(LOGGER_NAME).removeHandler(this)

    private fun diagnosticsFields(message: String?): Map<String, String>? {
        if (message?.startsWith("message flow ") != true) return null
        return message.split(' ').mapNotNull { token ->
            val separator = token.indexOf('=')
            if (separator > 0) token.substring(0, separator) to token.substring(separator + 1) else null
        }.toMap()
    }

    companion object {
        private const val LOGGER_NAME = "systems.zlink.framework.runtime.diagnostics.ZLinkMessageFlowTracer"
    }
}
